File: searchInUnknownLength.py
class Solution:
    def binarySearch(self,reader,low,high,target):
        while(low<=high):
            mid=low+(high-low)//2
            if reader.get(mid)==target:
                return mid
            elif target>reader.get(mid):
                low=mid+1
            else:
                high=mid-1
        return -1
    def search(self, reader, target):
        low=0
        high=1
        while reader.get(high)<target:
            low=high
            high=2*high
        return(self.binarySearch(reader,low,high,target))

#TC: O(log n)
#SC: O(1)

File: SearchInUnknownLength.java
// Time Complexity : O(logn)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

interface ArrayReader {
     public int get(int index);
 }

class SearchInUnknownLength {
    public int search(ArrayReader reader, int target) {
        int low = 0, high = 1;
        
        while(reader.get(high) < target) {
            low = high;
            high *= 2;
        }
        
        while(low <= high) {
            int mid = low + (high - low) / 2;
            
            int element = reader.get(mid);
            
            if(element == target) {
                return mid;
            } else if(target < element) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return -1;
    }
}